This is pulled from https://github.com/ungoogled-software/ungoogled-chromium/pull/3966

(Note: This is not part of the official ungoogled-chromium patch set)

--- a/components/web_package/BUILD.gn
+++ b/components/web_package/BUILD.gn
@@ -7,6 +7,7 @@
 source_set("web_package_headers") {
   visibility = [ ":*" ]
   sources = [
+    "input_reader.h",
     "signed_web_bundles/constants.h",
     "signed_web_bundles/identity_validator.h",
     "signed_web_bundles/integrity_block_parser.h",
@@ -52,8 +52,6 @@
     "//components/base32",
     "//components/cbor",
     "//components/web_package/mojom",
-    "//components/web_package/rust:web_package_rust_bindings",
-    "//components/web_package/signed_web_bundles/rust:signed_web_bundles_rust_bindings",
     "//mojo/public/cpp/bindings",
     "//net",
     "//services/network/public/mojom",
@@ -61,6 +62,7 @@
 
 static_library("web_package") {
   sources = [
+    "input_reader.cc",
     "signed_web_bundles/identity_validator.cc",
     "signed_web_bundles/signed_web_bundle_id.cc",
     "signed_web_bundles/signed_web_bundle_integrity_block.cc",
@@ -115,7 +114,6 @@
   deps = [
     ":mojom_mapped_types_headers",
     "//base",
-    "//components/web_package/signed_web_bundles/rust:signed_web_bundles_rust_bindings",
     "//crypto",
     "//mojo/public/cpp/bindings",
   ]
@@ -160,7 +158,6 @@
     "//base/test:test_support",
     "//components/cbor",
     "//components/web_package/rust:web_package_rust_unittests",
-    "//components/web_package/signed_web_bundles/rust:signed_web_bundles_rust_unittests",
     "//components/web_package/test_support",
     "//mojo/public/cpp/bindings",
     "//mojo/public/cpp/test_support:test_utils",
--- a/components/web_package/signed_web_bundles/ecdsa_p256_sha256_signature.cc
+++ b/components/web_package/signed_web_bundles/ecdsa_p256_sha256_signature.cc
@@ -9,7 +9,8 @@
 #include "base/strings/stringprintf.h"
 #include "base/types/expected.h"
 #include "components/web_package/signed_web_bundles/ecdsa_p256_public_key.h"
-#include "components/web_package/signed_web_bundles/rust/signed_web_bundles_rust.h"
+#include "crypto/keypair.h"
+#include "crypto/sign.h"
 
 namespace web_package {
 
@@ -50,8 +51,9 @@ EcdsaP256SHA256Signature::EcdsaP256SHA25
 [[nodiscard]] bool EcdsaP256SHA256Signature::Verify(
     base::span<const uint8_t> message,
     const EcdsaP256PublicKey& public_key) const {
-  return signed_web_bundles::rust::verify_ecdsa_p256_signature(
-      public_key.bytes(), bytes(), message);
+  auto key = crypto::keypair::PublicKey::FromEcP256Point(public_key.bytes());
+  return crypto::sign::Verify(crypto::sign::ECDSA_SHA256, *key, message,
+                              bytes());
 }
 
 }  // namespace web_package
--- a/components/web_package/signed_web_bundles/ed25519_signature.cc
+++ b/components/web_package/signed_web_bundles/ed25519_signature.cc
@@ -7,7 +7,6 @@
 #include <algorithm>
 
 #include "base/strings/stringprintf.h"
-#include "components/web_package/signed_web_bundles/rust/signed_web_bundles_rust.h"
 #include "third_party/boringssl/src/include/openssl/curve25519.h"
 
 namespace web_package {
@@ -43,8 +42,11 @@ Ed25519Signature::Ed25519Signature(std::
 [[nodiscard]] bool Ed25519Signature::Verify(
     base::span<const uint8_t> message,
     const Ed25519PublicKey& public_key) const {
-  return signed_web_bundles::rust::verify_ed25519_signature(public_key.bytes(),
-                                                            bytes(), message);
+  const std::array<uint8_t, ED25519_PUBLIC_KEY_LEN>& public_key_bytes =
+      public_key.bytes();
+  const std::array<uint8_t, ED25519_SIGNATURE_LEN>& signature_bytes = bytes();
+  return ED25519_verify(message.data(), message.size(), signature_bytes.data(),
+                        public_key_bytes.data());
 }
 
 }  // namespace web_package
--- a/components/web_package/signed_web_bundles/integrity_block_parser.cc
+++ b/components/web_package/signed_web_bundles/integrity_block_parser.cc
@@ -4,17 +4,19 @@
 
 #include "components/web_package/signed_web_bundles/integrity_block_parser.h"
 
+#include <algorithm>
 #include <string>
-#include <string_view>
 #include <vector>
 
-#include "base/check.h"
-#include "base/containers/span.h"
+#include "base/containers/map_util.h"
 #include "base/functional/bind.h"
 #include "base/strings/stringprintf.h"
+#include "base/strings/to_string.h"
 #include "base/types/expected.h"
 #include "base/types/expected_macros.h"
 #include "components/cbor/reader.h"
+#include "components/cbor/writer.h"
+#include "components/web_package/mojom/web_bundle_parser.mojom-forward.h"
 #include "components/web_package/mojom/web_bundle_parser.mojom.h"
 #include "components/web_package/signed_web_bundles/constants.h"
 #include "components/web_package/signed_web_bundles/ecdsa_p256_public_key.h"
@@ -22,56 +24,11 @@
 #include "components/web_package/signed_web_bundles/ed25519_public_key.h"
 #include "components/web_package/signed_web_bundles/ed25519_signature.h"
 #include "components/web_package/signed_web_bundles/integrity_block_attributes.h"
-#include "components/web_package/signed_web_bundles/rust/signed_web_bundles_rust.h"
 #include "components/web_package/signed_web_bundles/signed_web_bundle_id.h"
 #include "components/web_package/signed_web_bundles/types.h"
 
 namespace web_package {
 
-namespace {
-
-using SignatureType = signed_web_bundles::rust::SignatureType::Tag;
-using signed_web_bundles::rust::parse_integrity_block;
-
-base::expected<mojom::BundleIntegrityBlockSignatureStackEntryPtr, std::string>
-CreateSignatureStackEntry(
-    const signed_web_bundles::rust::SignatureStackEntry& entry) {
-  auto parsed_sig = mojom::BundleIntegrityBlockSignatureStackEntry::New();
-
-  switch (entry.signature_type.tag) {
-    case SignatureType::Ed25519: {
-      ASSIGN_OR_RETURN(auto public_key,
-                       Ed25519PublicKey::Create(entry.public_key.to_span()));
-      ASSIGN_OR_RETURN(auto signature,
-                       Ed25519Signature::Create(entry.signature.to_span()));
-      parsed_sig->signature_info = mojom::SignatureInfo::NewEd25519(
-          mojom::SignatureInfoEd25519::New(public_key, signature));
-      break;
-    }
-    case SignatureType::EcdsaP256SHA256: {
-      ASSIGN_OR_RETURN(auto public_key,
-                       EcdsaP256PublicKey::Create(entry.public_key.to_span()));
-      ASSIGN_OR_RETURN(auto signature, EcdsaP256SHA256Signature::Create(
-                                           entry.signature.to_span()));
-      parsed_sig->signature_info = mojom::SignatureInfo::NewEcdsaP256Sha256(
-          mojom::SignatureInfoEcdsaP256SHA256::New(public_key, signature));
-      break;
-    }
-    case SignatureType::Unknown: {
-      parsed_sig->signature_info =
-          mojom::SignatureInfo::NewUnknown(mojom::SignatureInfoUnknown::New());
-      break;
-    }
-  }
-
-  parsed_sig->attributes_cbor =
-      BinaryData(entry.attributes_cbor.begin(), entry.attributes_cbor.end());
-
-  return parsed_sig;
-}
-
-}  // namespace
-
 IntegrityBlockParser::IntegrityBlockParser(
     mojom::BundleDataSource& data_source,
     WebBundleParser::ParseIntegrityBlockCallback callback)
@@ -103,37 +60,103 @@ void IntegrityBlockParser::OnIntegrityBl
     RunErrorCallback("Error reading the integrity block.");
     return;
   }
+  // This structure is built during parsing and returned by the final callback
+  mojom::BundleIntegrityBlockPtr integrity_block =
+      mojom::BundleIntegrityBlock::New();
+
+  size_t consumed_bytes = 0;
+  cbor::Reader::DecoderError error;
+  cbor::Reader::Config config;
+  config.num_bytes_consumed = &consumed_bytes;
+  config.error_code_out = &error;
+
+  // When config `bytes_consumed` are assigned the CBOR parser reads only to the
+  // end of a structure. It parses only Integrity Block even if Web Bundle part
+  // starts just after.
+  std::optional<cbor::Value> value = cbor::Reader::Read(*data, config);
+  if (!value) {
+    RunErrorCallback("Error parsing integrity block as CBOR: " +
+                     std::string(cbor::Reader::ErrorCodeToString(error)));
+    return;
+  }
+  integrity_block->size = consumed_bytes;
+
+  if (!value->is_array()) {
+    RunErrorCallback("Integrity block is not a CBOR array.");
+    return;
+  }
 
-  const auto parse_res = parse_integrity_block(*data);
-  if (!parse_res.has_value()) {
-    const auto& error = parse_res.error();
-    RunErrorCallback(std::string(error.message.as_str()),
-                     error.is_version_error
-                         ? mojom::BundleParseErrorType::kVersionError
-                         : mojom::BundleParseErrorType::kFormatError);
+  const cbor::Value::ArrayValue& top_array = value->GetArray();
+  if (top_array.size() != kIntegrityBlockV2TopLevelArrayLength) {
+    RunErrorCallback(base::StringPrintf(
+        "Invalid integrity block array length: expected %u, got %zu.",
+        kIntegrityBlockV2TopLevelArrayLength, top_array.size()));
     return;
   }
 
-  const auto& parsed_ib = *parse_res;
-  auto integrity_block = mojom::BundleIntegrityBlock::New();
-  integrity_block->size = parsed_ib.size;
-
-  const std::string_view web_bundle_id_str =
-      parsed_ib.web_bundle_id.to_string_view();
-  RETURN_IF_ERROR(
-      SignedWebBundleId::Create(web_bundle_id_str),
-      [&](std::string error) { RunErrorCallback(std::move(error)); });
-
-  integrity_block->attributes =
-      IntegrityBlockAttributes(std::string(web_bundle_id_str),
-                               BinaryData(parsed_ib.attributes_cbor.begin(),
-                                          parsed_ib.attributes_cbor.end()));
+  // 1. Magic bytes
+  if (!top_array[0].is_bytestring() ||
+      !std::ranges::equal(top_array[0].GetBytestring(),
+                          kIntegrityBlockMagicBytes)) {
+    RunErrorCallback("Unexpected magic bytes.");
+    return;
+  }
+
+  // 2. Version
+  if (!top_array[1].is_bytestring() ||
+      !std::ranges::equal(top_array[1].GetBytestring(),
+                          kIntegrityBlockV2VersionBytes)) {
+    RunErrorCallback("Unexpected version bytes.",
+                     mojom::BundleParseErrorType::kVersionError);
+    return;
+  }
 
-  for (const auto& entry : parsed_ib.signature_stack) {
+  // 3. Attributes
+  if (!top_array[2].is_map()) {
+    RunErrorCallback("Integrity block attributes must be a map.");
+    return;
+  }
+  const cbor::Value::MapValue& attributes_map = top_array[2].GetMap();
+  const cbor::Value* web_bundle_id =
+      base::FindOrNull(attributes_map, cbor::Value(kWebBundleIdAttributeName));
+  if (!web_bundle_id || !web_bundle_id->is_string()) {
+    RunErrorCallback(
+        "`webBundleId` integrity block attribute is missing or malformed.");
+    return;
+  }
+
+  // Check if web_bundle_id is correct
+  RETURN_IF_ERROR(SignedWebBundleId::Create(web_bundle_id->GetString()),
+                  [&](const std::string& error) { RunErrorCallback(error); });
+
+  BinaryData attributes_cbor = *cbor::Writer::Write(top_array[2]);
+  integrity_block->attributes = IntegrityBlockAttributes(
+      web_bundle_id->GetString(), std::move(attributes_cbor));
+
+  // 4. Signature Stack
+  if (!top_array[3].is_array()) {
+    RunErrorCallback("Signature stack must be an array.");
+    return;
+  }
+  const cbor::Value::ArrayValue& signatures = top_array[3].GetArray();
+  if (signatures.empty()) {
+    RunErrorCallback(
+        "The signature stack must contain at least one signature.");
+    return;
+  }
+
+  for (const auto& signature_info_raw : signatures) {
     ASSIGN_OR_RETURN(
-        auto parsed_sig, CreateSignatureStackEntry(entry),
-        [&](std::string error) { RunErrorCallback(std::move(error)); });
-    integrity_block->signature_stack.push_back(std::move(parsed_sig));
+        auto parsed_signature_info, ParseSignatureInfo(signature_info_raw),
+        [&](const std::string& error) { this->RunErrorCallback(error); });
+
+    if (integrity_block->signature_stack.empty() &&
+        parsed_signature_info->signature_info->is_unknown()) {
+      RunErrorCallback("Unknown cipher type of the first signature.");
+      return;
+    }
+    integrity_block->signature_stack.push_back(
+        std::move(parsed_signature_info));
   }
 
   std::move(complete_callback_)
@@ -141,13 +164,74 @@ void IntegrityBlockParser::OnIntegrityBl
                           std::move(integrity_block), nullptr));
 }
 
+base::expected<mojom::BundleIntegrityBlockSignatureStackEntryPtr, std::string>
+IntegrityBlockParser::ParseSignatureInfo(
+    const cbor::Value& signature_info_raw) {
+  if (!signature_info_raw.is_array() ||
+      signature_info_raw.GetArray().size() != 2) {
+    return base::unexpected(
+        "Each signature stack entry must contain exactly two elements.");
+  }
+
+  const cbor::Value::ArrayValue& signature_info_array =
+      signature_info_raw.GetArray();
+  if (!signature_info_array[0].is_map() ||
+      !signature_info_array[1].is_bytestring()) {
+    return base::unexpected("Malformed signature stack entry.");
+  }
+
+  mojom::BundleIntegrityBlockSignatureStackEntryPtr parsed_signature_info =
+      mojom::BundleIntegrityBlockSignatureStackEntry::New();
+  parsed_signature_info->attributes_cbor =
+      *cbor::Writer::Write(signature_info_array[0]);
+
+  const cbor::Value::MapValue& signature_attributes =
+      signature_info_array[0].GetMap();
+  const BinaryData& signature_bytes = signature_info_array[1].GetBytestring();
+
+  const cbor::Value* ed25519_key = base::FindOrNull(
+      signature_attributes, cbor::Value(kEd25519PublicKeyAttributeName));
+  const cbor::Value* ecdsa_key = base::FindOrNull(
+      signature_attributes, cbor::Value(kEcdsaP256PublicKeyAttributeName));
+
+  if (ed25519_key && ecdsa_key) {
+    return base::unexpected("Multiple key types for one signature.");
+  } else if (ed25519_key && !ed25519_key->is_bytestring()) {
+    return base::unexpected("Invalid ED25519 key.");
+  } else if (ecdsa_key && !ecdsa_key->is_bytestring()) {
+    return base::unexpected("Invalid ECDSA key.");
+
+  } else if (ed25519_key && ed25519_key->is_bytestring()) {
+    ASSIGN_OR_RETURN(auto public_key,
+                     Ed25519PublicKey::Create(ed25519_key->GetBytestring()));
+    ASSIGN_OR_RETURN(auto signature, Ed25519Signature::Create(signature_bytes));
+    parsed_signature_info->signature_info = mojom::SignatureInfo::NewEd25519(
+        mojom::SignatureInfoEd25519::New(public_key, signature));
+
+  } else if (ecdsa_key && ecdsa_key->is_bytestring()) {
+    ASSIGN_OR_RETURN(auto public_key,
+                     EcdsaP256PublicKey::Create(ecdsa_key->GetBytestring()));
+    ASSIGN_OR_RETURN(auto signature,
+                     EcdsaP256SHA256Signature::Create(signature_bytes));
+    parsed_signature_info->signature_info =
+        mojom::SignatureInfo::NewEcdsaP256Sha256(
+            mojom::SignatureInfoEcdsaP256SHA256::New(public_key, signature));
+
+  } else {
+    parsed_signature_info->signature_info =
+        mojom::SignatureInfo::NewUnknown(mojom::SignatureInfoUnknown::New());
+  }
+
+  return parsed_signature_info;
+}
+
 void IntegrityBlockParser::RunErrorCallback(
-    std::string message,
+    const std::string& message,
     mojom::BundleParseErrorType error_type) {
   std::move(complete_callback_)
-      .Run(base::BindOnce(std::move(result_callback_), nullptr,
-                          mojom::BundleIntegrityBlockParseError::New(
-                              error_type, std::move(message))));
+      .Run(base::BindOnce(
+          std::move(result_callback_), nullptr,
+          mojom::BundleIntegrityBlockParseError::New(error_type, message)));
 }
 
 }  // namespace web_package
--- a/components/web_package/signed_web_bundles/integrity_block_parser.h
+++ b/components/web_package/signed_web_bundles/integrity_block_parser.h
@@ -5,13 +5,8 @@
 #ifndef COMPONENTS_WEB_PACKAGE_SIGNED_WEB_BUNDLES_INTEGRITY_BLOCK_PARSER_H_
 #define COMPONENTS_WEB_PACKAGE_SIGNED_WEB_BUNDLES_INTEGRITY_BLOCK_PARSER_H_
 
-#include <optional>
-#include <string>
-
-#include "base/compiler_specific.h"
-#include "base/memory/raw_ref.h"
-#include "base/memory/weak_ptr.h"
 #include "components/web_package/mojom/web_bundle_parser.mojom-forward.h"
+#include "components/web_package/signed_web_bundles/integrity_block_attributes.h"
 #include "components/web_package/signed_web_bundles/types.h"
 #include "components/web_package/web_bundle_parser.h"
 
@@ -20,7 +15,7 @@ namespace web_package {
 class IntegrityBlockParser : public WebBundleParser::WebBundleSectionParser {
  public:
   explicit IntegrityBlockParser(
-      mojom::BundleDataSource& data_source LIFETIME_BOUND,
+      mojom::BundleDataSource& data_source,
       WebBundleParser::ParseIntegrityBlockCallback callback);
 
   IntegrityBlockParser(const IntegrityBlockParser&) = delete;
@@ -35,7 +30,10 @@ class IntegrityBlockParser : public WebB
  private:
   void OnIntegrityBlockRead(const std::optional<BinaryData>& data);
 
-  void RunErrorCallback(std::string message,
+  base::expected<mojom::BundleIntegrityBlockSignatureStackEntryPtr, std::string>
+  ParseSignatureInfo(const cbor::Value& attributes_map);
+
+  void RunErrorCallback(const std::string& message,
                         mojom::BundleParseErrorType error_type =
                             mojom::BundleParseErrorType::kFormatError);
 
--- a/components/web_package/signed_web_bundles/rust/BUILD.gn
+++ b/components/web_package/signed_web_bundles/rust/BUILD.gn
@@ -10,7 +10,6 @@ rust_static_library("signed_web_bundles_
   sources = [
     "constants.rs",
     "integrity_block.rs",
-    "signature_verifier.rs",
     "types.rs",
     crate_root,
   ]
--- a/components/web_package/signed_web_bundles/signed_web_bundle_signature_verifier.cc
+++ b/components/web_package/signed_web_bundles/signed_web_bundle_signature_verifier.cc
@@ -20,14 +20,17 @@
 #include "base/task/thread_pool.h"
 #include "base/types/expected.h"
 #include "base/types/expected_macros.h"
+#include "components/cbor/values.h"
+#include "components/cbor/writer.h"
+#include "components/web_package/signed_web_bundles/constants.h"
 #include "components/web_package/signed_web_bundles/identity_validator.h"
 #include "components/web_package/signed_web_bundles/integrity_block_parser.h"
-#include "components/web_package/signed_web_bundles/rust/signed_web_bundles_rust.h"
 #include "components/web_package/signed_web_bundles/signed_web_bundle_id.h"
 #include "components/web_package/signed_web_bundles/signed_web_bundle_integrity_block.h"
 #include "components/web_package/signed_web_bundles/signed_web_bundle_signature_stack.h"
 #include "components/web_package/signed_web_bundles/signed_web_bundle_signature_stack_entry.h"
 #include "components/web_package/signed_web_bundles/signed_web_bundle_utils.h"
+#include "crypto/hash.h"
 #include "third_party/abseil-cpp/absl/functional/overload.h"
 #include "third_party/boringssl/src/include/openssl/sha.h"
 
@@ -37,10 +40,32 @@ namespace {
 
 std::vector<uint8_t> CreateIntegrityBlockCbor(
     const SignedWebBundleIntegrityBlock& integrity_block) {
-  auto ib_cbor_rust =
-      signed_web_bundles::rust::create_empty_integrity_block_cbor(
-          integrity_block.attributes_cbor());
-  return std::vector<uint8_t>(ib_cbor_rust.begin(), ib_cbor_rust.end());
+  // This function looks like it should use a cbor::Value::ArrayValue to build
+  // the outer array, but it can't: one of the elements is a pre-encoded CBOR
+  // value which we can't safely decode because it's untrusted input. This is
+  // obviously undesirable, because there's no guarantee at all that the
+  // pre-encoded value is actually valid CBOR.
+  //
+  // TODO(https://crbug.com/454485901): once there's a memory-safe CBOR parser,
+  // re-parse the pre-encoded value, then re-encode it here instead of
+  // constructing a CBOR array by hand so that we're guaranteed that the result
+  // of this function is valid.
+  std::vector<uint8_t> ib_cbor;
+
+  // 0x84 is the encoding byte for an array of length 4.
+  ib_cbor.push_back(0x84);
+  base::Extend(ib_cbor,
+               *cbor::Writer::Write(cbor::Value(kIntegrityBlockMagicBytes)));
+  base::Extend(ib_cbor, *cbor::Writer::Write(
+                            cbor::Value(kIntegrityBlockV2VersionBytes)));
+  // We cannot parse `attributes().cbor()` back to a cbor::Value as they
+  // technically represent untrusted input.
+  base::Extend(ib_cbor, integrity_block.attributes_cbor());
+  // Encode an empty signature array.
+  base::Extend(ib_cbor,
+               *cbor::Writer::Write(cbor::Value(cbor::Value::ArrayValue())));
+
+  return ib_cbor;
 }
 
 base::expected<void, SignedWebBundleSignatureVerifier::Error>
@@ -104,7 +129,7 @@ SignedWebBundleSignatureVerifier::Calcul
     return base::unexpected(base::File::ErrorToString(file.GetLastFileError()));
   }
 
-  signed_web_bundles::rust::Sha512Hasher hasher;
+  crypto::hash::Hasher hash(crypto::hash::kSha512);
 
   // Calculate the hash of the Signed Web Bundle excluding its integrity block.
   // The file might be too big to read it into memory all at once, which is why
@@ -121,14 +146,15 @@ SignedWebBundleSignatureVerifier::Calcul
           base::File::ErrorToString(file.GetLastFileError()));
     }
     data.resize(*bytes_read);
-    hasher.update(data);
+    hash.Update(data);
 
     if (!base::CheckAdd(offset, *bytes_read).AssignIfValid(&offset)) {
       return base::unexpected("The Signed Web Bundle is too large.");
     }
   }
 
-  SHA512Digest digest = std::move(hasher).finish();
+  SHA512Digest digest;
+  hash.Finish(digest);
   return digest;
 }
 
--- a/components/web_package/signed_web_bundles/signed_web_bundle_utils.cc
+++ b/components/web_package/signed_web_bundles/signed_web_bundle_utils.cc
@@ -5,18 +5,37 @@
 #include "components/web_package/signed_web_bundles/signed_web_bundle_utils.h"
 
 #include <cstdint>
-#include <vector>
 
-#include "components/web_package/signed_web_bundles/rust/signed_web_bundles_rust.h"
+#include "base/containers/extend.h"
+#include "base/containers/span.h"
+#include "base/containers/span_writer.h"
 
 namespace web_package {
 
+namespace {
+
+void AddItemToPayload(std::vector<uint8_t>& payload,
+                      base::span<const uint8_t> item) {
+  // Each item that is part of the payload is prefixed with its length encoded
+  // as a 64 bit unsigned integer.
+  std::array<uint8_t, sizeof(uint64_t)> length;
+  auto writer = base::SpanWriter(base::span(length));
+  CHECK(writer.WriteU64BigEndian(item.size()));
+
+  base::Extend(payload, base::span(length));
+  base::Extend(payload, item);
+}
+
+}  // namespace
+
 std::vector<uint8_t> CreateSignaturePayload(
     const SignedWebBundleSignatureData& data) {
-  auto payload_rust = signed_web_bundles::rust::create_signature_payload(
-      data.unsigned_web_bundle_hash, data.integrity_block_cbor,
-      data.attributes_cbor);
-  return std::vector<uint8_t>(payload_rust.begin(), payload_rust.end());
+  std::vector<uint8_t> payload;
+  AddItemToPayload(payload, data.unsigned_web_bundle_hash);
+  AddItemToPayload(payload, data.integrity_block_cbor);
+  AddItemToPayload(payload, data.attributes_cbor);
+
+  return payload;
 }
 
 }  // namespace web_package
--- a/third_party/blink/renderer/platform/BUILD.gn
+++ b/third_party/blink/renderer/platform/BUILD.gn
@@ -186,6 +186,18 @@ group("make_platform_generated") {
   ]
 }
 
+rust_static_library("font_format_check") {
+  allow_unsafe = true  # Needed for FFI that underpins the `cxx` crate.
+  crate_root = "fonts/opentype/format_check.rs"
+  sources = [ crate_root ]
+  cxx_bindings = [ crate_root ]
+  deps = [
+    "//third_party/rust/font_types/v0_12:lib",
+    "//third_party/rust/read_fonts/v0_43:lib",
+    "//third_party/rust/skrifa/v0_46:lib",
+  ]
+}
+
 rust_static_library("rustfft_ffi") {
   allow_unsafe = true  # Needed for FFI that underpins the `cxx` crate.
   crate_root = "audio/rustfft_ffi.rs"
@@ -1807,7 +1807,6 @@
     ":allow_discouraged_type",
     ":blink_platform_public_deps",
     ":platform_export",
-    "//build/rust/crubit",
     "//components/language_detection/core",
     "//gpu/command_buffer/client:raster_interface",
     "//media/capture:capture_lib",
@@ -1834,8 +1833,8 @@
     "//ui/native_theme/features",
   ]
   deps = [
+    ":font_format_check",
     ":rustfft_ffi",
-    "fonts:font_format_bindings",
     "fonts:ift_patcher_bindings",
     "//base:base_static",
     "//base/allocator:buildflags",
--- a/third_party/blink/renderer/platform/fonts/BUILD.gn
+++ b/third_party/blink/renderer/platform/fonts/BUILD.gn
@@ -1,23 +1,3 @@
-# Copyright 2026 The Chromium Authors
-# Use of this source code is governed by a BSD-style license that can be
-# found in the LICENSE file.
-
-import("//build/rust/rust_static_library.gni")
-
-rust_static_library("font_format") {
-  crate_root = "opentype/format_check.rs"
-  sources = [ crate_root ]
-  cpp_api_from_rust = {
-    target_name = "font_format_bindings"
-    cpp_namespace = "font_format"
-  }
-  deps = [
-    "//third_party/rust/font_types/v0_12:lib",
-    "//third_party/rust/read_fonts/v0_43:lib",
-    "//third_party/rust/skrifa/v0_46:lib",
-  ]
-}
-
 rust_static_library("ift_patcher_rs") {
   crate_root = "ift/ift_patcher.rs"
   sources = [ crate_root ]
--- a/third_party/blink/renderer/platform/fonts/opentype/font_format_check.cc
+++ b/third_party/blink/renderer/platform/fonts/opentype/font_format_check.cc
@@ -7,6 +7,7 @@
 #include "base/containers/span.h"
 #include "base/containers/span_rust.h"
 #include "base/numerics/byte_conversions.h"
+#include "third_party/blink/renderer/platform/fonts/opentype/format_check.rs.h"
 #include "third_party/blink/renderer/platform/runtime_enabled_features.h"
 #include "third_party/blink/renderer/platform/wtf/vector.h"
 #include "third_party/skia/include/core/SkTypeface.h"
@@ -14,37 +15,35 @@
 namespace blink {
 
 FontFormatCheck::FontFormatCheck(sk_sp<SkData> sk_data)
-    : format_info_(font_format::get_font_format_info(
+    : format_info_(font_format_check::get_font_format_info(
           base::SpanToRustSlice(sk_data->byteSpan()))) {}
 
-FontFormatCheck::~FontFormatCheck() = default;
-
 bool FontFormatCheck::IsVariableFont() const {
-  return font_format::is_variable(format_info_);
+  return font_format_check::is_variable(*format_info_);
 }
 
 bool FontFormatCheck::IsCbdtCblcColorFont() const {
-  return font_format::is_cbdt_cblc(format_info_);
+  return font_format_check::is_cbdt_cblc(*format_info_);
 }
 
 bool FontFormatCheck::IsEbdtEblcMonochromeFont() const {
-  return font_format::is_ebdt_eblc(format_info_);
+  return font_format_check::is_ebdt_eblc(*format_info_);
 }
 
 bool FontFormatCheck::IsColrCpalColorFontV0() const {
-  return font_format::is_colrv0(format_info_);
+  return font_format_check::is_colrv0(*format_info_);
 }
 
 bool FontFormatCheck::IsColrCpalColorFontV1() const {
-  return font_format::is_colrv1(format_info_);
+  return font_format_check::is_colrv1(*format_info_);
 }
 
 bool FontFormatCheck::IsSbixColorFont() const {
-  return font_format::is_sbix(format_info_);
+  return font_format_check::is_sbix(*format_info_);
 }
 
 bool FontFormatCheck::IsCff2OutlineFont() const {
-  return font_format::is_cff2(format_info_);
+  return font_format_check::is_cff2(*format_info_);
 }
 
 bool FontFormatCheck::IsVariableColrV0Font() const {
@@ -58,7 +57,7 @@ bool FontFormatCheck::IsColorFont() cons
 
 bool FontFormatCheck::IsAvar2Font() const {
   return RuntimeEnabledFeatures::FontFormatAvar2Enabled() &&
-         font_format::is_avar2(format_info_);
+         font_format_check::is_avar2(*format_info_);
 }
 
 FontFormatCheck::VariableFontSubType FontFormatCheck::ProbeVariableFont(
--- a/third_party/blink/renderer/platform/fonts/opentype/font_format_check.h
+++ b/third_party/blink/renderer/platform/fonts/opentype/font_format_check.h
@@ -5,7 +5,7 @@
 #ifndef THIRD_PARTY_BLINK_RENDERER_PLATFORM_FONTS_OPENTYPE_FONT_FORMAT_CHECK_H_
 #define THIRD_PARTY_BLINK_RENDERER_PLATFORM_FONTS_OPENTYPE_FONT_FORMAT_CHECK_H_
 
-#include "third_party/blink/renderer/platform/fonts/font_format.h"
+#include "third_party/blink/renderer/platform/fonts/opentype/format_check.rs.h"
 #include "third_party/blink/renderer/platform/platform_export.h"
 #include "third_party/blink/renderer/platform/wtf/allocator/allocator.h"
 #include "third_party/skia/include/core/SkData.h"
@@ -19,7 +19,7 @@ class PLATFORM_EXPORT FontFormatCheck {
 
  public:
   explicit FontFormatCheck(sk_sp<SkData>);
-  virtual ~FontFormatCheck();
+  virtual ~FontFormatCheck() = default;
   virtual bool IsVariableFont() const;
   virtual bool IsCbdtCblcColorFont() const;
   virtual bool IsEbdtEblcMonochromeFont() const;
@@ -46,7 +46,7 @@ class PLATFORM_EXPORT FontFormatCheck {
   enum class COLRVersion { kCOLRV0, kCOLRV1, kNoCOLR };
 
  private:
-  font_format::FontFormatInfo format_info_;
+  rust::Box<font_format_check::FontFormatInfo> format_info_;
 };
 
 }  // namespace blink
--- a/third_party/blink/renderer/platform/fonts/opentype/format_check.rs
+++ b/third_party/blink/renderer/platform/fonts/opentype/format_check.rs
@@ -23,7 +23,7 @@ pub struct FontFormatInfo {
     format_flags: Option<FontFormatFlags>,
 }
 
-pub fn get_font_format_info(font_bytes: &[u8]) -> FontFormatInfo {
+pub fn get_font_format_info(font_bytes: &[u8]) -> Box<FontFormatInfo> {
     let file_ref = make_font_ref_internal(font_bytes, 0);
 
     match file_ref {
@@ -32,11 +32,11 @@ pub fn get_font_format_info(font_bytes:
                 font.table_directory().table_records().iter().map(|e| e.tag()).collect();
             let color_version = get_colr_version(&font);
             let avar_version = get_avar_version(&font);
-            FontFormatInfo {
+            Box::new(FontFormatInfo {
                 format_flags: Some(FontFormatFlags { table_tags, color_version, avar_version }),
-            }
+            })
         }
-        _ => FontFormatInfo::default(),
+        _ => Box::new(FontFormatInfo::default()),
     }
 }
 
@@ -44,10 +44,10 @@ fn get_colr_version(font_ref: &FontRef)
     Some(font_ref.colr().ok()?.version())
 }
 
-pub fn is_colrv1(format_info: &FontFormatInfo) -> bool {
+fn is_colrv1(format_info: &FontFormatInfo) -> bool {
     matches!(&format_info.format_flags, Some(FontFormatFlags { color_version: Some(1), .. }),)
 }
-pub fn is_colrv0(format_info: &FontFormatInfo) -> bool {
+fn is_colrv0(format_info: &FontFormatInfo) -> bool {
     matches!(&format_info.format_flags, Some(FontFormatFlags { color_version: Some(0), .. }),)
 }
 
@@ -58,7 +58,7 @@
 
 // https://crbug.com/513446410: For safety we want to pass unknown avar2
 // versions through fontations, too.
-pub fn is_avar2(format_info: &FontFormatInfo) -> bool {
+fn is_avar2(format_info: &FontFormatInfo) -> bool {
     matches!(
         &format_info.format_flags,
         Some(FontFormatFlags {
@@ -69,22 +69,39 @@ fn has_tags(format_info: &FontFormatInfo
     }
 }
 
-pub fn is_variable(format_info: &FontFormatInfo) -> bool {
+fn is_variable(format_info: &FontFormatInfo) -> bool {
     has_tags(format_info, &[Tag::new(b"fvar")])
 }
 
-pub fn is_sbix(format_info: &FontFormatInfo) -> bool {
+fn is_sbix(format_info: &FontFormatInfo) -> bool {
     has_tags(format_info, &[Tag::new(b"sbix")])
 }
 
-pub fn is_cbdt_cblc(format_info: &FontFormatInfo) -> bool {
+fn is_cbdt_cblc(format_info: &FontFormatInfo) -> bool {
     has_tags(format_info, &[Tag::new(b"CBDT"), Tag::new(b"CBLC")])
 }
 
-pub fn is_ebdt_eblc(format_info: &FontFormatInfo) -> bool {
+fn is_ebdt_eblc(format_info: &FontFormatInfo) -> bool {
     has_tags(format_info, &[Tag::new(b"EBDT"), Tag::new(b"EBLC")])
 }
 
-pub fn is_cff2(format_info: &FontFormatInfo) -> bool {
+fn is_cff2(format_info: &FontFormatInfo) -> bool {
     has_tags(format_info, &[Tag::new(b"CFF2")])
 }
+
+#[cxx::bridge(namespace = "font_format_check")]
+pub mod ffi {
+    extern "Rust" {
+        type FontFormatInfo;
+
+        fn get_font_format_info(font_bytes: &[u8]) -> Box<FontFormatInfo>;
+        fn is_colrv1(format_info: &FontFormatInfo) -> bool;
+        fn is_colrv0(format_info: &FontFormatInfo) -> bool;
+        fn is_cbdt_cblc(format_info: &FontFormatInfo) -> bool;
+        fn is_ebdt_eblc(format_info: &FontFormatInfo) -> bool;
+        fn is_variable(format_info: &FontFormatInfo) -> bool;
+        fn is_sbix(format_info: &FontFormatInfo) -> bool;
+        fn is_cff2(format_info: &FontFormatInfo) -> bool;
+        fn is_avar2(format_info: &FontFormatInfo) -> bool;
+    }
+}
--- chromium-156.0.8078.4.orig/components/web_package/web_bundle_parser.cc	2026-09-30 01:57:21.000000000 +0200
+++ chromium-155.0.8059.5/components/web_package/web_bundle_parser.cc	2026-09-16 02:15:15.000000000 +0200
@@ -7,33 +7,175 @@
 #include <algorithm>
 #include <memory>
 #include <optional>
-#include <string>
 #include <string_view>
-#include <utility>
-#include <vector>
 
 #include "base/check.h"
 #include "base/compiler_specific.h"
-#include "base/containers/flat_map.h"
+#include "base/containers/span.h"
 #include "base/functional/bind.h"
 #include "base/functional/callback.h"
 #include "base/logging.h"
-#include "base/memory/raw_ref.h"
 #include "base/memory/weak_ptr.h"
 #include "base/notreached.h"
-#include "base/numerics/safe_conversions.h"
-#include "base/strings/strcat.h"
+#include "base/numerics/checked_math.h"
+#include "base/strings/string_number_conversions.h"
 #include "base/strings/string_util.h"
-#include "base/strings/string_view_util.h"
+#include "base/strings/stringprintf.h"
+#include "components/cbor/reader.h"
+#include "components/web_package/input_reader.h"
 #include "components/web_package/mojom/web_bundle_parser.mojom.h"
-#include "components/web_package/rust/web_package_rust.h"
 #include "components/web_package/signed_web_bundles/integrity_block_parser.h"
+#include "components/web_package/web_bundle_utils.h"
 #include "mojo/public/cpp/bindings/remote.h"
+#include "net/http/http_util.h"
 
 namespace web_package {
 
 namespace {
 
+// The number of bytes used to specify the length of the web bundle.
+// https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-trailing-length
+constexpr uint64_t kTrailingLengthNumBytes = 8;
+
+// The maximum size of the section-lengths CBOR item.
+constexpr uint64_t kMaxSectionLengthsCBORSize = 8192;
+
+// The maximum size of a metadata section allowed in this implementation.
+constexpr uint64_t kMaxMetadataSectionSize = 1 * 1024 * 1024;
+
+// The maximum size of the response header CBOR.
+constexpr uint64_t kMaxResponseHeaderLength = 512 * 1024;
+
+// The initial buffer size for reading an item from the response section.
+constexpr uint64_t kInitialBufferSizeForResponse = 4096;
+
+// The first byte of WebBundle format >=b2 (Array of length 5).
+constexpr uint8_t kBundleHeadByte = 0x85;
+// The first byte of WebBundle format b1 (Array of length 6).
+constexpr uint8_t kBundleB1HeadByte = 0x86;
+
+// CBOR of the magic string "🌐📦".
+// MetadataParser::ParseMagicBytes() checks the first byte (0x85) and this.
+//
+// The first 10 bytes of the web bundle format are:
+//   85                             -- Array of length 5
+//      48                          -- Byte string of length 8
+//         F0 9F 8C 90 F0 9F 93 A6  -- "🌐📦" in UTF-8
+const uint8_t kBundleMagicBytes[] = {
+    0x48, 0xF0, 0x9F, 0x8C, 0x90, 0xF0, 0x9F, 0x93, 0xA6,
+};
+
+// CBOR of the version string "b2\0\0".
+//   44               -- Byte string of length 4
+//       62 32 00 00  -- "b2\0\0"
+const uint8_t kVersionB2MagicBytes[] = {
+    0x44, 0x62, 0x32, 0x00, 0x00,
+};
+// CBOR of the version string "b1\0\0".
+//   44               -- Byte string of length 4
+//       62 31 00 00  -- "b1\0\0"
+const uint8_t kVersionB1MagicBytes[] = {
+    0x44, 0x62, 0x31, 0x00, 0x00,
+};
+
+// Section names.
+constexpr char kCriticalSection[] = "critical";
+constexpr char kIndexSection[] = "index";
+constexpr char kPrimarySection[] = "primary";
+constexpr char kResponsesSection[] = "responses";
+
+// A list of (section-name, length) pairs.
+using SectionLengths = std::vector<std::pair<std::string, uint64_t>>;
+
+// A map from section name to (offset, length) pair.
+using SectionOffsets = std::map<std::string, std::pair<uint64_t, uint64_t>>;
+
+bool IsMetadataSection(const std::string& name) {
+  return (name == kCriticalSection || name == kIndexSection ||
+          name == kPrimarySection);
+}
+
+// Parses a `section-lengths` CBOR item.
+// https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-bundle-sections
+//   section-lengths = [* (section-name: tstr, length: uint) ]
+std::optional<SectionLengths> ParseSectionLengths(
+    base::span<const uint8_t> data) {
+  cbor::Reader::DecoderError error;
+  std::optional<cbor::Value> value = cbor::Reader::Read(data, &error);
+  if (!value.has_value() || !value->is_array()) {
+    return std::nullopt;
+  }
+
+  const cbor::Value::ArrayValue& array = value->GetArray();
+  if (array.size() % 2 != 0) {
+    return std::nullopt;
+  }
+
+  SectionLengths result;
+  for (size_t i = 0; i < array.size(); i += 2) {
+    if (!array[i].is_string() || !array[i + 1].is_unsigned()) {
+      return std::nullopt;
+    }
+    result.emplace_back(array[i].GetString(), array[i + 1].GetUnsigned());
+  }
+  return result;
+}
+
+struct ParsedHeaders {
+  base::flat_map<std::string, std::string> headers;
+  base::flat_map<std::string, std::string> pseudos;
+};
+
+// https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-responses
+//   headers = {* bstr => bstr}
+std::optional<ParsedHeaders> ConvertCBORValueToHeaders(
+    const cbor::Value& headers_value) {
+  // |headers_value| of headers must be a map.
+  if (!headers_value.is_map()) {
+    return std::nullopt;
+  }
+
+  ParsedHeaders result;
+
+  for (const auto& item : headers_value.GetMap()) {
+    if (!item.first.is_bytestring() || !item.second.is_bytestring()) {
+      return std::nullopt;
+    }
+    std::string_view name = item.first.GetBytestringAsString();
+    std::string_view value = item.second.GetBytestringAsString();
+
+    // If name contains any upper-case or non-ASCII characters, return an error.
+    // This matches the requirement in Section 8.1.2 of [RFC7540].
+    if (!base::IsStringASCII(name) ||
+        std::ranges::any_of(name, base::IsAsciiUpper<char>)) {
+      return std::nullopt;
+    }
+
+    if (!name.empty() && name[0] == ':') {
+      // pseudos[name] must not exist, because CBOR maps cannot contain
+      // duplicate keys. This is ensured by cbor::Reader.
+      DCHECK(!result.pseudos.contains(name));
+      result.pseudos.insert(
+          std::make_pair(std::string(name), std::string(value)));
+      continue;
+    }
+
+    // Both name and value must be valid.
+    if (!net::HttpUtil::IsValidHeaderName(name) ||
+        !net::HttpUtil::IsValidHeaderValue(value)) {
+      return std::nullopt;
+    }
+
+    // headers[name] must not exist, because CBOR maps cannot contain duplicate
+    // keys. This is ensured by cbor::Reader.
+    DCHECK(!result.headers.contains(name));
+
+    result.headers.insert(
+        std::make_pair(std::string(name), std::string(value)));
+  }
+
+  return result;
+}
 
 GURL ParseExchangeURL(std::string_view str, const GURL& base_url) {
   DCHECK(base_url.is_empty() || base_url.is_valid());
@@ -128,9 +270,7 @@
       RunErrorCallback("Error reading bundle length.");
       return;
     }
-    const uint64_t unsigned_file_length =
-        base::checked_cast<uint64_t>(file_length);
-    if (unsigned_file_length < rust::TRAILING_LENGTH_NUM_BYTES) {
+    if (static_cast<uint64_t>(file_length) < kTrailingLengthNumBytes) {
       RunErrorCallback("Error reading bundle length.");
       return;
     }
@@ -138,10 +278,9 @@
     // Read the last 8 bytes of the file that correspond to the trailing length
     // field of the web bundle.
     data_source_->get()->Read(
-        unsigned_file_length - rust::TRAILING_LENGTH_NUM_BYTES,
-        rust::TRAILING_LENGTH_NUM_BYTES,
+        file_length - kTrailingLengthNumBytes, kTrailingLengthNumBytes,
         base::BindOnce(&MetadataParser::ParseWebBundleLength,
-                       weak_factory_.GetWeakPtr(), unsigned_file_length));
+                       weak_factory_.GetWeakPtr(), file_length));
   }
 
   void ParseWebBundleLength(const uint64_t file_length,
@@ -156,20 +295,30 @@
     // the start of the bundle, instead of assuming that the start of the file
     // is also the start of the bundle. This allows the bundle to be appended to
     // another format such as a generic self-extracting executable."
-    // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-trailing-length
-    auto bundle_offset = rust::parse_trailing_length(*data, file_length);
-    if (!bundle_offset.has_value()) {
-      RunErrorCallback(bundle_offset.error().message);
+    // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#section-4.1.1-3
+    InputReader input(*data);
+    uint64_t web_bundle_length;
+    if (!input.ReadBigEndian(&web_bundle_length)) {
+      RunErrorCallback("Error reading bundle length.");
+      return;
+    }
+
+    if (web_bundle_length > file_length) {
+      RunErrorCallback("Invalid bundle length.");
       return;
     }
-    ReadMagicBytes(*bundle_offset);
+    const uint64_t web_bundle_offset = file_length - web_bundle_length;
+    ReadMagicBytes(web_bundle_offset);
   }
 
   void ReadMagicBytes(const uint64_t offset_in_stream) {
     // First, we will parse the CBOR header of the top level array (1-byte),
     // `magic`, `version`, and the CBOR header of `section-lengths`.
+    const uint64_t length = 1 + sizeof(kBundleMagicBytes) +
+                            sizeof(kVersionB2MagicBytes) +
+                            kMaxCBORItemHeaderSize;
     data_source_->get()->Read(
-        offset_in_stream, rust::INITIAL_BUNDLE_HEADER_BUFFER_SIZE,
+        offset_in_stream, length,
         base::BindOnce(&MetadataParser::ParseMagicBytes,
                        weak_factory_.GetWeakPtr(), offset_in_stream));
   }
@@ -182,23 +331,83 @@
       return;
     }
 
-    auto res = rust::parse_magic_and_version(*data, offset_in_stream);
-    if (!res.has_value()) {
-      const auto& error = res.error();
-      RunErrorCallback(error.message,
-                       error.is_version_error
-                           ? mojom::BundleParseErrorType::kVersionError
-                           : mojom::BundleParseErrorType::kFormatError);
+    InputReader input(*data);
+
+    // Read the first byte denoting a CBOR array size. It must be equal to 0x85
+    // (5).
+    const auto array_size = input.ReadByte();
+    if (!array_size) {
+      RunErrorCallback("Missing CBOR array size byte.");
       return;
     }
 
-    uint64_t next_offset = res->next_read_offset;
-    uint64_t next_length = res->next_read_length;
+    // Let kBundleB1HeadByte pass this check, to report custom error message for
+    // b1 bundles.
+    if (*array_size != kBundleHeadByte && *array_size != kBundleB1HeadByte) {
+      RunErrorCallback("Wrong magic bytes.");
+      return;
+    }
+
+    // Check the magic bytes "48 F0 9F 8C 90 F0 9F 93 A6".
+    const auto magic = input.ReadBytes(sizeof(kBundleMagicBytes));
+    if (!magic || !std::ranges::equal(*magic, kBundleMagicBytes)) {
+      RunErrorCallback("Wrong magic bytes.");
+      return;
+    }
+
+    // Let version be the result of reading 5 bytes from stream.
+    const auto version = input.ReadBytes(sizeof(kVersionB2MagicBytes));
+    if (!version) {
+      RunErrorCallback("Cannot read version bytes.");
+      return;
+    }
+    if (!std::ranges::equal(*version, kVersionB2MagicBytes)) {
+      const char* message;
+      if (std::ranges::equal(*version, kVersionB1MagicBytes)) {
+        message =
+            "Bundle format version is 'b1' which is no longer supported."
+            " Currently supported version is: 'b2'";
+      } else {
+        message =
+            "Version error: bundle format does not correspond to the specifed "
+            "version. Currently supported version is: 'b2'";
+      }
+      RunErrorCallback(message, mojom::BundleParseErrorType::kVersionError);
+      return;
+    }
+    if (*array_size != kBundleHeadByte) {
+      RunErrorCallback("Wrong CBOR array size of the top-level structure");
+      return;
+    }
+
+    const auto section_lengths_length =
+        input.ReadCBORHeader(CBORType::kByteString);
+    if (!section_lengths_length) {
+      RunErrorCallback("Cannot parse the size of section-lengths.");
+      return;
+    }
+
+    // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-top-level-structure
+    // "The section-lengths array is embedded in a byte string to facilitate
+    // reading it from a network. This byte string MUST be less than 8192
+    // (8*1024) bytes long, and parsers MUST NOT load any data from a
+    // section-lengths item longer than this."
+    if (*section_lengths_length >= kMaxSectionLengthsCBORSize) {
+      RunErrorCallback(
+          "The section-lengths CBOR must be smaller than 8192 bytes.");
+      return;
+    }
+
+    // In the next step, we will parse the content of `section-lengths`,
+    // and the CBOR header of `sections`.
+    const uint64_t length = *section_lengths_length + kMaxCBORItemHeaderSize;
+
+    offset_in_stream += input.CurrentOffset();
     data_source_->get()->Read(
-        next_offset, next_length,
+        offset_in_stream, length,
         base::BindOnce(&MetadataParser::ParseBundleHeader,
-                       weak_factory_.GetWeakPtr(), next_offset,
-                       res->section_lengths_len));
+                       weak_factory_.GetWeakPtr(), offset_in_stream,
+                       *section_lengths_length));
   }
 
   void ParseBundleHeader(uint64_t offset_in_stream,
@@ -208,39 +417,109 @@
       RunErrorCallback("Error reading bundle header.");
       return;
     }
+    InputReader input(*data);
+
+    // webbundle = [
+    //    magic: h'F0 9F 8C 90 F0 9F 93 A6',
+    //    version: bytes .size 4,
+    //    section-lengths: bytes .cbor section-lengths,  <==== here
+    //    sections: [* any ],
+    //    length: bytes .size 8,  ; Big-endian number of bytes in the bundle.
+    // ]
+    const auto section_lengths_bytes = input.ReadBytes(section_lengths_length);
+    if (!section_lengths_bytes) {
+      RunErrorCallback("Cannot read section-lengths.");
+      return;
+    }
+    // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-bundle-sections
+    //   section-lengths = [* (section-name: tstr, length: uint) ]
+    const auto section_lengths = ParseSectionLengths(*section_lengths_bytes);
+    if (!section_lengths) {
+      RunErrorCallback("Cannot parse section-lengths.");
+      return;
+    }
+
+    // webbundle = [
+    //    magic: h'F0 9F 8C 90 F0 9F 93 A6',
+    //    version: bytes .size 4,
+    //    section-lengths: bytes .cbor section-lengths,
+    //    sections: [* any ],  <==== here
+    //    length: bytes .size 8,  ; Big-endian number of bytes in the bundle.
+    // ]
+    const auto num_sections = input.ReadCBORHeader(CBORType::kArray);
+    if (!num_sections) {
+      RunErrorCallback("Cannot parse the number of sections.");
+      return;
+    }
 
-    auto res = rust::parse_bundle_header(*data, section_lengths_length,
-                                         offset_in_stream);
-    if (!res.has_value()) {
-      RunErrorCallback(res.error().message);
+    // "The sections array contains the sections' content. The length of this
+    // array MUST be exactly half the length of the section-lengths array, and
+    // parsers MUST NOT load any data if that is not the case."
+    if (*num_sections != section_lengths->size()) {
+      RunErrorCallback("Unexpected number of sections.");
       return;
     }
 
-    responses_section_offset_ = res->responses_offset;
-    responses_section_length_ = res->responses_length;
-    metadata_sections_to_read_ = std::move(res->metadata_sections);
+    const uint64_t sections_start = offset_in_stream + input.CurrentOffset();
+    uint64_t current_offset = sections_start;
+
+    // Convert |section_lengths| to |section_offsets_|.
+    for (const auto& pair : *section_lengths) {
+      const std::string& name = pair.first;
+      const uint64_t length = pair.second;
+      bool added = section_offsets_
+                       .insert(std::make_pair(
+                           name, std::make_pair(current_offset, length)))
+                       .second;
+      if (!added) {
+        RunErrorCallback("Duplicated section.");
+        return;
+      }
+
+      if (!base::CheckAdd(current_offset, length)
+               .AssignIfValid(&current_offset)) {
+        RunErrorCallback("Integer overflow calculating section offsets.");
+        return;
+      }
+    }
+
+    // "The "responses" section MUST appear after the other three sections
+    // defined here, and parsers MUST NOT load any data if that is not the
+    // case."
+    if (section_lengths->empty() ||
+        section_lengths->back().first != kResponsesSection) {
+      RunErrorCallback("Responses section is not the last in section-lengths.");
+      return;
+    }
 
+    // Initialize |metadata_|.
     metadata_ = mojom::BundleMetadata::New();
     metadata_->version = mojom::BundleFormatVersion::kB2;
 
-    ReadMetadataSections(/*section_index=*/0);
+    ReadMetadataSections(section_offsets_.begin());
   }
 
   // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-bundle-sections
-  void ReadMetadataSections(size_t section_index) {
-    if (section_index < metadata_sections_to_read_.size()) {
-      const auto& section = metadata_sections_to_read_[section_index];
-      if (section.length > rust::MAX_METADATA_SECTION_SIZE) {
+  void ReadMetadataSections(SectionOffsets::const_iterator section_iter) {
+    for (; section_iter != section_offsets_.end(); ++section_iter) {
+      const auto& name = section_iter->first;
+      if (!IsMetadataSection(name)) {
+        continue;
+      }
+      const uint64_t section_offset = section_iter->second.first;
+      const uint64_t section_length = section_iter->second.second;
+      if (section_length > kMaxMetadataSectionSize) {
         RunErrorCallback(
             "Metadata sections larger than 1MB are not supported.");
         return;
       }
 
       data_source_->get()->Read(
-          section.offset, section.length,
+          section_offset, section_length,
           base::BindOnce(&MetadataParser::ParseMetadataSection,
-                         weak_factory_.GetWeakPtr(), section_index,
-                         section.length));
+                         weak_factory_.GetWeakPtr(), section_iter,
+                         section_length));
+      // This loop will be resumed by ParseMetadataSection().
       return;
     }
 
@@ -253,7 +532,7 @@
     RunSuccessCallback();
   }
 
-  void ParseMetadataSection(size_t section_index,
+  void ParseMetadataSection(SectionOffsets::const_iterator section_iter,
                             uint64_t expected_data_length,
                             const std::optional<std::vector<uint8_t>>& data) {
     if (!data || data->size() != expected_data_length) {
@@ -261,56 +540,152 @@
       return;
     }
 
-    const auto& name = metadata_sections_to_read_[section_index].name;
-    if (name == rust::INDEX_SECTION) {
-      auto res = rust::parse_index_section(*data, responses_section_offset_,
-                                           responses_section_length_);
-      if (!res.has_value()) {
-        RunErrorCallback(res.error().message);
+    // Parse the section contents as a CBOR item.
+    cbor::Reader::DecoderError error;
+    std::optional<cbor::Value> section_value =
+        cbor::Reader::Read(*data, &error);
+    if (!section_value) {
+      RunErrorCallback(std::string("Error parsing section contents as CBOR: ") +
+                       cbor::Reader::ErrorCodeToString(error));
+      return;
+    }
+
+    const auto& name = section_iter->first;
+    // Note: Parse*Section() delete |this| on failure.
+    if (name == kIndexSection) {
+      if (!ParseIndexSection(*section_value)) {
         return;
       }
-      std::vector<std::pair<GURL, mojom::BundleResponseLocationPtr>> requests;
-      requests.reserve(res->size());
-      for (const auto& entry : *res) {
-        GURL parsed_url = ParseExchangeURL(entry.url, base_url_);
-        if (!parsed_url.is_valid()) {
-          std::string message = base::StrCat({"Index section: exchange URL \"",
-                                              entry.url, "\" is not valid."});
-          if (base_url_.is_empty()) {
-            message += " (Relative URLs are not allowed in this context.)";
-          }
-          RunErrorCallback(message);
-          return;
-        }
-        requests.emplace_back(
-            std::move(parsed_url),
-            mojom::BundleResponseLocation::New(entry.offset, entry.length));
-      }
-      metadata_->requests =
-          base::flat_map<GURL, mojom::BundleResponseLocationPtr>(
-              std::move(requests));
-    } else if (name == rust::CRITICAL_SECTION) {
-      if (auto res = rust::parse_critical_section(*data); !res.has_value()) {
-        RunErrorCallback(res.error().message);
+    } else if (name == kCriticalSection) {
+      if (!ParseCriticalSection(*section_value)) {
         return;
       }
-    } else if (name == rust::PRIMARY_SECTION) {
-      auto res = rust::parse_primary_section(*data);
-      if (!res.has_value()) {
-        RunErrorCallback(res.error().message);
+    } else if (name == kPrimarySection) {
+      if (!ParsePrimarySection(*section_value)) {
         return;
       }
-      GURL parsed_url = ParseExchangeURL(*res, base_url_);
-      if (!parsed_url.is_valid()) {
-        RunErrorCallback("Primary URL is not a valid exchange URL.");
-        return;
-      }
-      metadata_->primary_url = std::move(parsed_url);
     } else {
       NOTREACHED();
     }
     // Read the next metadata section.
-    ReadMetadataSections(section_index + 1);
+    ReadMetadataSections(++section_iter);
+  }
+
+  // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-the-index-section
+  // The index section has the following structure:
+  //   index = {* whatwg-url => [ location-in-responses ] }
+  //   location-in-responses = (offset: uint, length: uint)
+  bool ParseIndexSection(const cbor::Value& section_value) {
+    // |section_value| of index section must be a map.
+    if (!section_value.is_map()) {
+      RunErrorCallback("Index section must be a map.");
+      return false;
+    }
+
+    base::flat_map<GURL, mojom::BundleResponseLocationPtr> requests;
+
+    auto responses_section = section_offsets_.find(kResponsesSection);
+    CHECK(responses_section != section_offsets_.end());
+    const uint64_t responses_section_offset = responses_section->second.first;
+    const uint64_t responses_section_length = responses_section->second.second;
+
+    // For each (url, responses) entry in the index map.
+    for (const auto& item : section_value.GetMap()) {
+      if (!item.first.is_string()) {
+        RunErrorCallback("Index section: key must be a string.");
+        return false;
+      }
+      if (!item.second.is_array()) {
+        RunErrorCallback("Index section: value must be an array.");
+        return false;
+      }
+      const std::string& url = item.first.GetString();
+      const cbor::Value::ArrayValue& responses_array = item.second.GetArray();
+
+      GURL parsed_url = ParseExchangeURL(url, base_url_);
+
+      if (!parsed_url.is_valid()) {
+        std::string message = base::StringPrintf(
+            "Index section: exchange URL \"%s\" is not valid.", url.c_str());
+        if (base_url_.is_empty()) {
+          message += " (Relative URLs are not allowed in this context.)";
+        }
+        RunErrorCallback(message);
+        return false;
+      }
+
+      if (responses_array.size() != 2) {
+        RunErrorCallback(
+            "Index section: the size of a response array per URL should be "
+            "exactly 2.");
+        return false;
+      }
+      if (!responses_array[0].is_unsigned() ||
+          !responses_array[1].is_unsigned()) {
+        RunErrorCallback(
+            "Index section: offset and length values must be unsigned.");
+        return false;
+      }
+      uint64_t offset = responses_array[0].GetUnsigned();
+      uint64_t length = responses_array[1].GetUnsigned();
+
+      uint64_t response_end;
+      if (!base::CheckAdd(offset, length).AssignIfValid(&response_end) ||
+          response_end > responses_section_length) {
+        RunErrorCallback("Index section: response out of range.");
+        return false;
+      }
+      uint64_t offset_within_stream = responses_section_offset + offset;
+
+      requests.insert(std::make_pair(
+          parsed_url,
+          mojom::BundleResponseLocation::New(offset_within_stream, length)));
+    }
+
+    metadata_->requests = std::move(requests);
+    return true;
+  }
+
+  // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#critical-section
+  //   critical = [*tstr]
+  bool ParseCriticalSection(const cbor::Value& section_value) {
+    if (!section_value.is_array()) {
+      RunErrorCallback("Critical section must be an array.");
+      return false;
+    }
+    // "If the client has not implemented a section named by one of the items in
+    // this list, the client MUST fail to parse the bundle as a whole."
+    for (const cbor::Value& elem : section_value.GetArray()) {
+      if (!elem.is_string()) {
+        RunErrorCallback("Non-string element in the critical section.");
+        return false;
+      }
+      const auto& section_name = elem.GetString();
+      if (!IsMetadataSection(section_name) &&
+          section_name != kResponsesSection) {
+        RunErrorCallback("Unknown critical section.");
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // https://github.com/WICG/webpackage/blob/main/extensions/primary-section.md
+  //  primary = whatwg-url
+  bool ParsePrimarySection(const cbor::Value& section_value) {
+    if (!section_value.is_string()) {
+      RunErrorCallback("Primary section must be a string.");
+      return false;
+    }
+
+    GURL parsed_url = ParseExchangeURL(section_value.GetString(), base_url_);
+
+    if (!parsed_url.is_valid()) {
+      RunErrorCallback("Primary URL is not a valid exchange URL.");
+      return false;
+    }
+    metadata_->primary_url = std::move(parsed_url);
+    return true;
   }
 
   void RunSuccessCallback() {
@@ -319,14 +694,12 @@
                             nullptr));
   }
 
-  void RunErrorCallback(std::string_view message,
+  void RunErrorCallback(const std::string& message,
                         mojom::BundleParseErrorType error_type =
                             mojom::BundleParseErrorType::kFormatError) {
     DLOG(ERROR) << "Parsing web bundle error: " << message;
-    // Avoids an extra temporary std::string allocation in Mojo's New().
-    auto err = mojom::BundleMetadataParseError::New();
-    err->type = error_type;
-    err->message = message;
+    mojom::BundleMetadataParseErrorPtr err =
+        mojom::BundleMetadataParseError::New(error_type, message);
     std::move(complete_callback_)
         .Run(base::BindOnce(std::move(result_callback_), nullptr,
                             std::move(err)));
@@ -337,9 +710,7 @@
   std::optional<uint64_t> start_reading_offset_;
   ParseMetadataCallback result_callback_;
   ParsingCompleteCallback complete_callback_;
-  rs_std::Vec<rust::SectionOffsetEntry> metadata_sections_to_read_;
-  uint64_t responses_section_offset_ = 0;
-  uint64_t responses_section_length_ = 0;
+  SectionOffsets section_offsets_;
   mojom::BundleMetadataPtr metadata_;
   base::WeakPtrFactory<MetadataParser> weak_factory_{this};
 };
@@ -373,7 +744,7 @@
       override {
     CHECK(!result_callback_.is_null());
     complete_callback_ = std::move(callback);
-    StartWithBufferSize(rust::INITIAL_BUFFER_SIZE_FOR_RESPONSE);
+    StartWithBufferSize(kInitialBufferSizeForResponse);
   }
 
  private:
@@ -384,36 +755,115 @@
         base::BindOnce(&ResponseParser::ParseResponseHeader,
                        weak_factory_.GetWeakPtr(), length));
   }
+  // https://www.ietf.org/archive/id/draft-ietf-wpack-bundled-responses-01.html#name-responses
+  //   responses = [*response]
+  //   response = [headers: bstr .cbor headers, payload: bstr]
+  //   headers = {* bstr => bstr}
   void ParseResponseHeader(uint64_t expected_data_length,
                            const std::optional<std::vector<uint8_t>>& data) {
     if (!data || data->size() != expected_data_length) {
       RunErrorCallback("Error reading response header.");
       return;
     }
+    InputReader input(*data);
+
+    // |response| must be an array of length 2 (headers and payload).
+    auto num_elements = input.ReadCBORHeader(CBORType::kArray);
+    if (!num_elements || *num_elements != 2) {
+      RunErrorCallback("Array size of response must be 2.");
+      return;
+    }
 
-    auto parse_res =
-        rust::parse_response(*data, response_offset_, response_length_);
-    if (!parse_res.has_value()) {
-      RunErrorCallback(parse_res.error().message);
+    auto header_length = input.ReadCBORHeader(CBORType::kByteString);
+    if (!header_length) {
+      RunErrorCallback("Cannot parse response header length.");
       return;
     }
-    if (parse_res->needs_more_data) {
+
+    // "The length of the headers byte string in a response MUST be less than
+    // 524288 (512*1024) bytes, and recipients MUST fail to load a response with
+    // longer headers"
+    if (*header_length >= kMaxResponseHeaderLength) {
+      RunErrorCallback("Response header is too big.");
+      return;
+    }
+
+    // If we don't have enough data for the headers and the CBOR header of the
+    // payload, re-read with a larger buffer size.
+    const uint64_t required_buffer_size = std::min(
+        input.CurrentOffset() + *header_length + kMaxCBORItemHeaderSize,
+        response_length_);
+    if (data->size() < required_buffer_size) {
       DVLOG(1) << "Re-reading response header with a buffer of size "
-               << parse_res->required_buffer_size;
-      StartWithBufferSize(parse_res->required_buffer_size);
+               << required_buffer_size;
+      StartWithBufferSize(required_buffer_size);
+      return;
+    }
+
+    // Parse headers.
+    auto headers_bytes = input.ReadBytes(*header_length);
+    if (!headers_bytes) {
+      RunErrorCallback("Cannot read response headers.");
+      return;
+    }
+    cbor::Reader::DecoderError error;
+    std::optional<cbor::Value> headers_value =
+        cbor::Reader::Read(*headers_bytes, &error);
+    if (!headers_value) {
+      RunErrorCallback("Cannot parse response headers.");
+      return;
+    }
+
+    auto parsed_headers = ConvertCBORValueToHeaders(*headers_value);
+    if (!parsed_headers) {
+      RunErrorCallback("Cannot parse response headers.");
+      return;
+    }
+
+    // "Each response's headers MUST include a :status pseudo-header with
+    // exactly 3 ASCII decimal digits and MUST NOT include any other
+    // pseudo-headers."
+    const auto pseudo_status = parsed_headers->pseudos.find(":status");
+    if (parsed_headers->pseudos.size() != 1 ||
+        pseudo_status == parsed_headers->pseudos.end()) {
+      RunErrorCallback(
+          "Response headers map must have exactly one pseudo-header, :status.");
+      return;
+    }
+    int status;
+    const auto& status_str = pseudo_status->second;
+    if (status_str.size() != 3 ||
+        !std::ranges::all_of(status_str, base::IsAsciiDigit<char>) ||
+        !base::StringToInt(status_str, &status)) {
+      RunErrorCallback(":status must be 3 ASCII decimal digits.");
+      return;
+    }
+
+    // Parse payload.
+    auto payload_length = input.ReadCBORHeader(CBORType::kByteString);
+    if (!payload_length) {
+      RunErrorCallback("Cannot parse response payload length.");
+      return;
+    }
+
+    // "If a response's payload is not empty, its headers MUST include a
+    // Content-Type header (Section 8.3 of [I-D.ietf-httpbis-semantics])."
+    if (*payload_length > 0 &&
+        !parsed_headers->headers.contains("content-type")) {
+      RunErrorCallback("Non-empty response must have a content-type header.");
+      return;
+    }
+
+    if (input.CurrentOffset() + *payload_length != response_length_) {
+      RunErrorCallback("Unexpected payload length.");
       return;
     }
 
     mojom::BundleResponsePtr response = mojom::BundleResponse::New();
-    response->response_code = parse_res->response_code;
-    response->response_headers = base::MakeFlatMap<std::string, std::string>(
-        parse_res->headers, {}, [](const rust::HeaderEntry& header) {
-          return std::make_pair(
-              std::string(base::as_string_view(header.name.to_span())),
-              std::string(base::as_string_view(header.value.to_span())));
-        });
-    response->payload_offset = parse_res->payload_offset;
-    response->payload_length = parse_res->payload_length;
+    response->response_code = status;
+    response->response_headers = std::move(parsed_headers->headers);
+    response->payload_offset = response_offset_ + input.CurrentOffset();
+    response->payload_length = *payload_length;
     RunSuccessCallback(std::move(response));
   }
 
@@ -423,15 +873,13 @@
                             nullptr));
   }
 
-  void RunErrorCallback(const std::string_view message,
+  void RunErrorCallback(const std::string& message,
                         mojom::BundleParseErrorType error_type =
                             mojom::BundleParseErrorType::kFormatError) {
-    auto err = mojom::BundleResponseParseError::New();
-    err->type = error_type;
-    err->message = message;
     std::move(complete_callback_)
-        .Run(base::BindOnce(std::move(result_callback_), nullptr,
-                            std::move(err)));
+        .Run(base::BindOnce(
+            std::move(result_callback_), nullptr,
+            mojom::BundleResponseParseError::New(error_type, message)));
   }
 
   const raw_ref<mojo::Remote<mojom::BundleDataSource>> data_source_;
--- /dev/null
+++ chromium-155.0.8059.5/components/web_package/input_reader.cc	1970-01-01 01:00:00.000000000 +0100
@@ -0,0 +1,193 @@
+// Copyright 2022 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "components/web_package/input_reader.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "base/numerics/checked_math.h"
+#include "base/strings/string_util.h"
+#include "base/strings/string_view_util.h"
+#include "components/cbor/constants.h"
+#include "components/cbor/values.h"
+
+namespace web_package {
+
+namespace {
+
+// This array must be kept in sync with the `CBORType` enum.
+constexpr std::array kAcceptedCBORTypes = {
+    // clang-format off
+    CBORType::kUnsignedInt,
+    CBORType::kNegativeInt,
+    CBORType::kByteString,
+    CBORType::kTextString,
+    CBORType::kArray,
+    CBORType::kMap,
+    CBORType::kSimpleValue,
+    // clang-format on
+};
+
+std::optional<int64_t> DecodeValueToNegative(uint64_t value) {
+  auto negative_value = -base::CheckedNumeric<int64_t>(value) - 1;
+  if (!negative_value.IsValid()) {
+    return std::nullopt;
+  }
+  return static_cast<int64_t>(negative_value.ValueOrDie());
+}
+
+std::optional<int64_t> DecodeValueToUnsigned(uint64_t value) {
+  auto unsigned_value = base::CheckedNumeric<int64_t>(value);
+  if (!unsigned_value.IsValid()) {
+    return std::nullopt;
+  }
+  return static_cast<int64_t>(unsigned_value.ValueOrDie());
+}
+
+}  // namespace
+
+InputReader::InputReader(base::span<const uint8_t> buf) : buf_(buf) {}
+
+InputReader::~InputReader() = default;
+
+std::optional<uint8_t> InputReader::ReadByte() {
+  uint8_t b;
+  if (!buf_.ReadU8BigEndian(b)) {
+    return std::nullopt;
+  }
+  return {b};
+}
+
+std::optional<base::span<const uint8_t>> InputReader::ReadBytes(size_t n) {
+  return buf_.Read(n);
+}
+
+std::optional<std::string_view> InputReader::ReadString(size_t n) {
+  auto bytes = buf_.Read(n);
+  if (!bytes) {
+    return std::nullopt;
+  }
+  if (!base::IsStringUTF8(base::as_string_view(*bytes))) {
+    return std::nullopt;
+  }
+  return base::as_string_view(*bytes);
+}
+
+std::optional<uint64_t> InputReader::ReadCBORHeader(CBORType expected_type) {
+  auto pair = ReadTypeAndArgument();
+  if (!pair || pair->first != expected_type) {
+    return std::nullopt;
+  }
+  return pair->second;
+}
+
+std::optional<CBORHeader> InputReader::ReadCBORHeader() {
+  auto pair = ReadTypeAndArgument();
+  if (!pair) {
+    return std::nullopt;
+  }
+
+  const auto& [type, additional_info] = *pair;
+  switch (type) {
+    case CBORType::kSimpleValue: {
+      using SimpleValue = cbor::Value::SimpleValue;
+      if (additional_info == std::to_underlying(SimpleValue::TRUE_VALUE)) {
+        return {{true}};
+      } else if (additional_info ==
+                 std::to_underlying(SimpleValue::FALSE_VALUE)) {
+        return {{false}};
+      } else {
+        return std::nullopt;
+      }
+    }
+    case CBORType::kUnsignedInt:
+    case CBORType::kNegativeInt: {
+      std::optional<int64_t> value =
+          type == CBORType::kUnsignedInt
+              ? DecodeValueToUnsigned(additional_info)
+              : DecodeValueToNegative(additional_info);
+      if (!value) {
+        return std::nullopt;
+      }
+      return {{*value}};
+    }
+    case CBORType::kByteString:
+    case CBORType::kTextString: {
+      using StringInfo = CBORHeader::StringInfo;
+      using StringType = CBORHeader::StringInfo::StringType;
+
+      return {{StringInfo{.type = type == CBORType::kByteString
+                                      ? StringType::kByteString
+                                      : StringType::kTextString,
+                          .byte_length = additional_info}}};
+    }
+    case CBORType::kArray:
+    case CBORType::kMap: {
+      using ContainerInfo = CBORHeader::ContainerInfo;
+      using ContainerType = CBORHeader::ContainerInfo::ContainerType;
+
+      return {{ContainerInfo{.type = type == CBORType::kArray
+                                         ? ContainerType::kArray
+                                         : ContainerType::kMap,
+                             .size = additional_info}}};
+    }
+  }
+}
+
+// https://datatracker.ietf.org/doc/html/rfc8949.html#section-3
+std::optional<std::pair<CBORType, uint64_t>>
+InputReader::ReadTypeAndArgument() {
+  std::optional<uint8_t> first_byte = ReadByte();
+  if (!first_byte) {
+    return std::nullopt;
+  }
+
+  // There are more CBOR types in the standard than we accept. To avoid mishits
+  // during `static_cast<CBORType>`, it's safer to validate the type beforehand.
+  uint8_t type_byte = (*first_byte & cbor::constants::kMajorTypeMask) >>
+                      cbor::constants::kMajorTypeBitShift;
+  if (!std::ranges::contains(kAcceptedCBORTypes, type_byte,
+                             &std::to_underlying<CBORType>)) {
+    return std::nullopt;
+  }
+
+  CBORType type = static_cast<CBORType>(type_byte);
+  uint8_t b = *first_byte & cbor::constants::kAdditionalInformationMask;
+
+  if (b <= 23) {
+    return std::make_pair(type, b);
+  }
+  if (b == 24) {
+    auto content = ReadByte();
+    if (!content || *content < 24) {
+      return std::nullopt;
+    }
+    return std::make_pair(type, *content);
+  }
+  if (b == 25) {
+    uint16_t content;
+    if (!ReadBigEndian(&content) || content >> 8 == 0) {
+      return std::nullopt;
+    }
+    return std::make_pair(type, content);
+  }
+  if (b == 26) {
+    uint32_t content;
+    if (!ReadBigEndian(&content) || content >> 16 == 0) {
+      return std::nullopt;
+    }
+    return std::make_pair(type, content);
+  }
+  if (b == 27) {
+    uint64_t content;
+    if (!ReadBigEndian(&content) || content >> 32 == 0) {
+      return std::nullopt;
+    }
+    return std::make_pair(type, content);
+  }
+  return std::nullopt;
+}
+
+}  // namespace web_package
--- /dev/null
+++ chromium-155.0.8059.5/components/web_package/input_reader.h	1970-01-01 01:00:00.000000000 +0100
@@ -0,0 +1,113 @@
+// Copyright 2022 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef COMPONENTS_WEB_PACKAGE_INPUT_READER_H_
+#define COMPONENTS_WEB_PACKAGE_INPUT_READER_H_
+
+#include <optional>
+#include <string_view>
+#include <variant>
+
+#include "base/containers/span.h"
+#include "base/containers/span_reader.h"
+#include "base/memory/stack_allocated.h"
+#include "base/types/id_type.h"
+
+namespace web_package {
+
+// https://datatracker.ietf.org/doc/html/rfc8949.html#section-3.1
+enum class CBORType {
+  kUnsignedInt = 0,
+  kNegativeInt = 1,
+  kByteString = 2,
+  kTextString = 3,
+  kArray = 4,
+  kMap = 5,
+  // kTag = 6,
+  kSimpleValue = 7,
+  // kFloatValue = 7,
+};
+
+struct CBORHeader {
+  struct StringInfo {
+    enum class StringType {
+      kByteString,
+      kTextString,
+    } type;
+    uint64_t byte_length;
+  };
+  struct ContainerInfo {
+    enum class ContainerType {
+      kArray,
+      kMap,
+    } type;
+    uint64_t size;
+  };
+
+  const std::variant<bool, int64_t, StringInfo, ContainerInfo> data;
+};
+
+// The maximum length of the CBOR item header (type and argument).
+// https://datatracker.ietf.org/doc/html/rfc8949.html#section-3
+// When the additional information (the low-order 5 bits of the first
+// byte) is 27, the argument's value is held in the following 8 bytes.
+constexpr uint64_t kMaxCBORItemHeaderSize = 9;
+
+// A utility class for reading various values from input buffer.
+class InputReader {
+  STACK_ALLOCATED();
+
+ public:
+  explicit InputReader(base::span<const uint8_t> buf);
+
+  InputReader(const InputReader&) = delete;
+  InputReader& operator=(const InputReader&) = delete;
+
+  ~InputReader();
+
+  size_t CurrentOffset() const { return buf_.num_read(); }
+  size_t Size() const { return buf_.remaining(); }
+
+  std::optional<uint8_t> ReadByte();
+
+  template <typename T>
+    requires(std::is_integral_v<T> && std::is_unsigned_v<T>)
+  bool ReadBigEndian(T* out) {
+    if constexpr (sizeof(T) == 1) {
+      return buf_.ReadU8BigEndian(*out);
+    } else if constexpr (sizeof(T) == 2) {
+      return buf_.ReadU16BigEndian(*out);
+    } else if constexpr (sizeof(T) == 4) {
+      return buf_.ReadU32BigEndian(*out);
+    } else {
+      static_assert(sizeof(T) == 8);
+      return buf_.ReadU64BigEndian(*out);
+    }
+  }
+
+  std::optional<base::span<const uint8_t>> ReadBytes(size_t n);
+
+  std::optional<std::string_view> ReadString(size_t n);
+
+  // Parses the type and argument of a CBOR item from the input head. If parsed
+  // successfully and the type matches `expected_type`, returns the argument.
+  // Otherwise returns nullopt.
+  std::optional<uint64_t> ReadCBORHeader(CBORType expected_type);
+
+  // Parses the type and argument of a CBOR item from the input head. If parsed
+  // successfully, returns type and:
+  //  * value for kUnsignedInt/kNegativeInt;
+  //  * value_size for kTextString/kByteString/kMap/kArray.
+  // Otherwise returns nullopt.
+  std::optional<CBORHeader> ReadCBORHeader();
+
+ private:
+  std::optional<std::pair<CBORType, uint64_t>> ReadTypeAndArgument();
+
+  base::SpanReader<const uint8_t> buf_;
+};
+
+}  // namespace web_package
+
+#endif  // COMPONENTS_WEB_PACKAGE_INPUT_READER_H_
